x-pack/plugins/infra/public/components/waffle/index.tsx

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { findDOMNode } from 'react-dom';
+import styled from 'styled-components';
+import { AutoSizer } from '../auto_sizer';
+import { GroupOfGroups } from './group_of_groups';
+import { GroupOfNodes } from './group_of_nodes';
+
+import {
+  isWaffleMapGroupWithGroups,
+  isWaffleMapGroupWithNodes,
+} from '../../containers/libs/type_guards';
+import { InfraWaffleData, InfraWaffleMapGroup, InfraWaffleOptions } from '../../lib/lib';
+import { applyWaffleMapLayout } from './lib/apply_wafflemap_layout';
+import { Scroll } from './scroll';
+
+interface WafleMapProps {
+  options: InfraWaffleOptions;
+  map: InfraWaffleData;
+}
+
+const initialState = {
+  hasScroll: false,
+};
+
+type WaffleMapState = Readonly<typeof initialState>;
+
+export class Waffle extends React.Component<WafleMapProps, WaffleMapState> {
+  public readonly state: WaffleMapState = initialState;
+  private outer: any;
+  private inner: any;
+  private scrollBarTimer: any;
+
+  public render() {
+    return (
+      <AutoSizer bounds>
+        {({ measureRef, bounds: { width = 0, height = 0 } }) => {
+          const { hasScroll } = this.state;
+          const { map } = this.props;
+          const groupsWithLayout = applyWaffleMapLayout(map, width, height);
+          return (
+            <Container innerRef={(el: any) => measureRef(el)}>
+              <Scroll hasScroll={hasScroll} innerRef={(el: any) => (this.outer = el)}>
+                <MapContainer innerRef={(el: any) => (this.inner = el)}>
+                  {groupsWithLayout.map(this.renderGroup)}
+                </MapContainer>
+              </Scroll>
+            </Container>
+          );
+        }}
+      </AutoSizer>
+    );
+  }
+
+  public componentWillMount() {
+    const check = () => {
+      this.scrollBarTimer = setTimeout(() => {
+        const el = findDOMNode(this.outer);
+        if (el instanceof Element) {
+          const hasScroll = el.scrollWidth > el.clientWidth || el.scrollHeight > el.clientHeight;
+          this.setState({ hasScroll }, check);
+        }
+      }, 500);
+    };
+    check();
```

Comment:
This seems to be the main source of the rendering churn. I'm not sure what the purpose is :thinking: With the simplified component structure above, could we get rid of this? Then we wouldn't need most of the private members and state of this component either.

Reply:
It's checking for the scroll bars.